src/xlogging.c

```diff
@@ -18,8 +38,123 @@ void xlogging_set_log_function(LOGGER_LOG log_function)
+        HANDLE process = GetCurrentProcess();
+
+        SYMBOL_INFO_EXTENDED symbolExtended;
+        SYMBOL_INFO* symbol = &symbolExtended.symbol;
+
+        IMAGEHLP_LINE64 line;
+
+        symbol->MaxNameLen = TRACE_MAX_SYMBOL_SIZE;
+        symbol->SizeOfStruct = sizeof(SYMBOL_INFO);
+
+        for (uint32_t j = 0; j < numberOfFrames; j++)
+        {
+            DWORD64 address = (DWORD64)(stack[j]);
+            DWORD displacement = 0;
+
+            if (!SymFromAddr(process, address, NULL, symbol))
+            {
+                (void)strncat(destination, "SymFromAddr failed\n", destinationSize - destinationPos);
+            }
+            else
+            {
+                line.SizeOfStruct = sizeof(IMAGEHLP_LINE64);
+                if (SymGetLineFromAddr64(process, address, &displacement, &line))
+                {
+                    int snprintfResult = snprintf(resultLine, sizeof(resultLine), "!%s %s:%" PRIu32 "\n", symbol->Name, line.FileName, line.LineNumber);
+                    if (!(
+                        (snprintfResult >= 0) && /*the returned value is nonnegative [...]*/
+                        (snprintfResult < sizeof(resultLine)) /*[...] and less than n.*/
+                        ))
+                    {
+                        (void)strncat(destination, "snprintf failed\n", destinationSize - destinationPos);
+                    }
+                    else
+                    {
+                        (void)strncat(destination, resultLine, destinationSize - destinationPos);
+                    }
+                }
+                else
+                {
+                    int snprintfResult = snprintf(resultLine, sizeof(resultLine), "!%s Address 0x%" PRIX64 "\n", symbol->Name, line.Address);
+                    if (!(
+                        (snprintfResult >= 0) && /*the returned value is nonnegative [...]*/
+                        (snprintfResult < sizeof(resultLine)) /*[...] and less than n.*/
+                        ))
+                    {
+                        (void)strncat(destination, "snprintf failed\n", destinationSize - destinationPos);
+                    }
+                    else
+                    {
+                        (void)strncat(destination, resultLine, destinationSize - destinationPos);
+                    }
+                }
+            }
+            destinationPos = strlen(destination);
+        }
+
+        ReleaseSRWLockExclusive(&lockOverSymCalls);
+        
+    }
+    return result;
+}
+#endif
+#endif
+
 LOGGER_LOG xlogging_get_log_function(void)
 {
+#if defined _MSC_VER
+#ifdef LOGERROR_CAPTURES_STACK_TRACES
+    LONG state;
+    while ((state = InterlockedCompareExchange(&doSymInit, 1, 0)) != 2)
+    {
+        if (state == 0)
+        {
+            (void)SymInitialize(GetCurrentProcess(), NULL, TRUE);
+            (void)InterlockedExchange(&doSymInit, 2);
+        }
+    }
+#endif
+#endif
     return global_log_function;
 }
```

Comment:
Frankly, this says we should add an xlogging_init/deinit.
Just so we don't piggy back the getter.

I don't think it would be hard to add. What says you?

Reply:
Me says "when time comes". 

---
In reply to: [276826626](https://github.com/Azure/azure-c-shared-utility/pull/302#discussion_r276826626) [](ancestors = 276826626)